cats/cats-sql/src/main/kotlin/com/netflix/spinnaker/cats/sql/cache/SqlCache.kt

```diff
@@ -649,123 +683,240 @@ class SqlCache(
+          val relType = relAgent.substringBefore(delimiter = ":", missingDelimiterValue = "")
+
+          rels.value.forEach { relId ->
+            val idHash = getIdHash(cacheData.id)
+            val relIdHash = getIdHash(relId)
+
+            val fwdKey = "$idHash|$relIdHash"
+            val revKey = "$relIdHash|$idHash"
+            currentIds.add(fwdKey)
+            currentIds.add(revKey)
+
+            result.relationshipCount.incrementAndGet()
+
+            if (!oldFwdIds.containsKey(fwdKey)) {
+              newFwdRelEntries.add(
+                RelEntry(
+                  uuid = null,
+                  id_hash = idHash,
+                  id = cacheData.id,
+                  rel_id_hash = relIdHash,
+                  rel_id = relId,
+                  rel_agent_hash = relAgentHash,
+                  rel_agent = relAgent,
+                  rel_type = relType,
+                  last_updated = null
+                )
+              )
+            }
 
-              if (!oldRevIds.containsKey(revKey)) {
-                newRevRelIds.add(revKey)
-              }
+            if (!oldRevIds.containsKey(revKey)) {
+              newRevTypeToRelEntries.getOrPut(relType) { mutableListOf() }
+                .add(
+                  RelEntry(
+                    uuid = null,
+                    id_hash = relIdHash,
+                    id = relId,
+                    rel_id_hash = idHash,
+                    rel_id = cacheData.id,
+                    rel_agent_hash = relAgentHash,
+                    rel_agent = relAgent,
+                    rel_type = type,
+                    last_updated = null
+                  )
+                )
             }
+          }
         }
       }
 
-    newFwdRelPointers.forEach { (relType, pointers) ->
-      val now = clock.millis()
-      var ulid = ULID().nextValue()
+    val now = clock.millis()
+    var ulid = ULID().nextValue()
 
-      pointers.chunked(dynamicConfigService.getConfig(Int::class.java, "sql.cache.write-batch-size", 100)) { chunk ->
+    newFwdRelEntries.chunked(
+      dynamicConfigService.getConfig(
+        Int::class.java,
+        "sql.cache.write-batch-size",
+        100
+      )
+    ) { chunk ->
+      try {
+        val insert = jooq.insertInto(
+          table(sqlNames.relTableName(type)),
+          field("uuid"),
+          field("id_hash"),
+          field("id"),
+          field("rel_id_hash"),
+          field("rel_id"),
+          field("rel_agent_hash"),
+          field("rel_agent"),
+          field("rel_type"),
+          field("last_updated")
+        )
+
+        insert.apply {
+          chunk.forEach {
+            values(
+              ulid.toString(),
+              it.id_hash,
+              it.id,
+              it.rel_id_hash,
+              it.rel_id,
+              it.rel_agent_hash,
+              it.rel_agent,
+              it.rel_type,
+              now
+            )
+            ulid = ULID().nextMonotonicValue(ulid)
```

Comment:
I think this is pre-existing, but I believe this is creating a new `ULID` object for each iteration of the loop (which from the implementation looks to also create a new `SecureRandom` for each iteration; this likely creates a lot of unecessary garbage.

I'd suggest pulling the `ULID` at least out of the loop, and potentially into an instance variable (if it's thread-safe).